include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h

```diff
@@ -47,11 +47,19 @@ class __reduce_mid_work_group_kernel;
 template <typename... _Name>
 class __reduce_kernel;
 
+// Storage helper since _Tp may not have a default constructor.
+template <typename _Tp>
+union __storage
```

Comment:
If we are going to lift this type definition out, we probably need to rename it as well.

Reply:
Additionally if we are going to move list this type definition out, we may cover the case when we have array of elements too.

Reply:
I'm not sure about the array of elements, perhaps that reaches too far beyond the scope of this PR, but maybe something like  `__delayed_ctor_storage`?  

I think we need something which describes its purpose.

Reply:
`__optional_ctor_storage` ?

Reply:
I suppose the first is the array case Sergey was referring to,  I'd be fine with leaving that one out for now to limit the scope of the PR if it makes it significantly more complicated.

Reply:
I propose to make additional changes with it in some separate PR.

Reply:
Sure, for this PR, lets just rename it, we can unify, etc. in a separate PR.
My vote is for `__lazy_ctor_storage` because I think optional advertises more functionality than is provided here.

Reply:
Thanks for this discussion. I agree that larger changes are outside the scope of this PR and change the naming to `__lazy_ctor_storage`.